src/picongpu/include/algorithms/KinEnergy.hpp

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2017 Heiko Burau, Axel Huebl
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "simulation_defines.hpp"
+#include "algorithms/Gamma.hpp"
+
+namespace picongpu
+{
+
+using namespace PMacc;
+
+template<typename precisionType = float_X>
+struct KinEnergy
```

Comment:
This algorithm is agnostic to the weighting factor. It can be used both for real particles *and* macro particles.